File: languages/l/lua/lambda-core.lua
-- _lc postfix to prevent conflict
-- Bool value
local true_lc = function(x)
	return function(y)
		return x
	end
end
local false_lc = function(x)
	return function(y)
		return y
	end
end

-- Boolean Operations
local and_lc = function(a)
	return function(b)
		return a(b)(false_lc)
	end
end

local or_lc = function(a)
	return function(b)
		return a(true_lc)(b)
	end
end

local not_lc = function(a)
	return a(false_lc)(true_lc)
end

-- Church Numerals
local zero = function(f)
	return function(x)
		return x
	end
end
local one = function(f)
	return function(x)
		return f(x)
	end
end
local succ = function(n)
	return function(f)
		return function(x)
			return f(n(f)(x))
		end
	end
end
-- This is ugly
local pred = function(n)
	return function(f)
		return function(x)
			return n(function(g)
				return function(h)
					return h(g(f))
				end
			end)(function(u)
				return x
			end)(function(u)
				return u
			end)
		end
	end
end
-- a slightly better version of pred
local pred_v2 = function(n)
	return function(f)
		return function(x)
			local helper = function(g)
				return function(h)
					return h(g(f))
				end
			end
			return n(helper)(function(_)
				return x
			end)(function(u)
				return u
			end)
		end
	end
end
-- Helper functions
local read_bool = function(b)
	return b(true)(false)
end

local read_church = function(n)
	return n(function(x)
		return x + 1
	end)(0)
end

-- Test Cases
print(read_bool(true_lc)) -- true
print(read_bool(false_lc)) -- false
print(read_bool(not_lc(true_lc))) -- false
print(read_bool(not_lc(false_lc))) -- true
print(read_bool(and_lc(true_lc)(true_lc))) -- true
print(read_bool(and_lc(true_lc)(false_lc))) -- false
print(read_bool(and_lc(false_lc)(true_lc))) -- false
print(read_bool(and_lc(false_lc)(false_lc))) -- false
print(read_bool(or_lc(true_lc)(true_lc))) -- true
print(read_bool(or_lc(true_lc)(false_lc))) -- true
print(read_bool(or_lc(false_lc)(true_lc))) -- true
print(read_bool(or_lc(false_lc)(false_lc))) -- false

print(read_church(zero)) -- 0
print(read_church(succ(zero))) -- 1
print(read_church(succ(one))) -- 2
print(read_church(succ(succ(one)))) -- 3
print(read_church(pred(one))) -- 0
print(read_church(pred_v2(one))) -- 0
print(read_church(pred(succ(one)))) -- 1
print(read_church(pred_v2(succ(one)))) -- 1
print(read_church(pred(succ(succ(one))))) -- 2
print(read_church(pred_v2(succ(succ(one))))) -- 2

File: languages/k/kotlin/lambda-core.kt
fun interface T : (T) -> T {}
// LOGIC    
val TRUE: T
   	get() = T {x: T -> T {y: T -> x}}
val FALSE: T
   	get() = T {x: T -> T {y: T -> y}}
val NOT: T 
   	get() = T {x: T -> x(FALSE)(TRUE)}
val AND: T
   	get() = T {x: T -> T {y: T -> x(y)(FALSE)}}
val OR: T
   	get() = T {x: T -> T {y: T -> x(TRUE)(y)}}

// CHURCH NUMERALS
val ZERO: T
	get() = T{f: T -> T {x: T -> x}}
val SUCC: T
	get() = T {n: T -> T {f: T -> T {x: T -> f(n(f)(x))}}}
val PRED: T
	get() = T {n: T -> T {f: T -> T {x: T -> n(T {g: T -> T {h: T -> h(g(f))}})(T {u -> x})(T {u -> u})}}}

// HELPERS - not pure lambda calculus
fun readBool(t: T) {
	var res: Boolean? = null
    val _true: T = T {x -> 
       	res = true
       	x
    }
    val _false: T = T {x -> 
        res = false
        x
    }
    t(_true)(_false)({t: T -> t})
    println(res)
}

fun readChurchNumber(t: T) {
    var res: Int = 0
    
    val plusOne: T = T {x: T ->
        res++
        x
    }
    
    t(plusOne)({t:T -> t})
    println(res)
}

// -------------------------------------------------

// EXAMPLES
fun main() {
    println("LOGIC")
    println("----------------")
    println("TRUE/FALSE")
    readBool(TRUE) // true
    readBool(FALSE) // false

    println("NOT")
    readBool(NOT(TRUE)) // false
    readBool(NOT(FALSE)) //true
    
    println("AND")
    readBool(AND(FALSE)(FALSE)) // false
    readBool(AND(FALSE)(TRUE)) // false
    readBool(AND(TRUE)(FALSE)) // false
    readBool(AND(TRUE)(TRUE)) // true
    
    println("OR")
    readBool(OR(FALSE)(FALSE)) // false
    readBool(OR(FALSE)(TRUE)) // true
    readBool(OR(TRUE)(FALSE)) // true
    readBool(OR(TRUE)(TRUE)) // true
    
    println("CHURCH NUMERALS")
    println("---------------")
    println("ZERO/SUCC")
    readChurchNumber(ZERO) // 0
    readChurchNumber(SUCC(ZERO)) // 1
    readChurchNumber(SUCC(SUCC(ZERO))) // 2
    
    println("PRED")
    val THREE = SUCC(SUCC(SUCC(ZERO))) // 3
    readChurchNumber(PRED(THREE)) // 2
    readChurchNumber(PRED(PRED(THREE))) // 1 
    readChurchNumber(PRED(PRED(PRED(THREE)))) // 0
}